audio-echo/app/src/main/java/com/google/sample/echo/MainActivity.java

```diff
@@ -120,11 +138,36 @@ private void queryNativeAudioParameters() {
         nativeSampleFormat ="";    //TODO: find a way to get the native audio format
     }
     private void updateNativeAudioUI() {
-        status_view.setText("nativeSampleRate    = " + nativeSampleRate + "\n" +
-                "nativeSampleBufSize = " + nativeSampleBufSize + "\n" +
-                "nativeSampleFormat  = " + nativeSampleFormat);
+            status_view.setText("nativeSampleRate    = " + nativeSampleRate + "\n" +
+                    "nativeSampleBufSize = " + nativeSampleBufSize + "\n" +
+                    "nativeSampleFormat  = " + nativeSampleFormat);
+    }
 
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+                                           @NonNull int[] grantResults) {
+        /*
+         * if any permission failed, the sample could not play
+         */
+        if (AUDIO_ECHO_REQUEST != requestCode) {
+            super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+            return;
+        }
+
+        if (grantResults.length != 1  ||
+            grantResults[0] != PackageManager.PERMISSION_GRANTED) {
+            status_view.setText("Error: failed to get user permission for RECORD_AUDIO");
```

Comment:
Let's make the error more explicit: "Error: RECORD_AUDIO permission was denied." or similar.